### PR TITLE
Stabler expit and logit derivatives

### PR DESCRIPTION
**Pull request by j-towns**
You get NAN when you run
```python
from jax.scipy.special import expit
from jax import grad

grad(expit)(-90.)
```
on GPU.

The gradient really should be zero there, not NAN. On CPU you get zero - which is correct, but I believe that's only because of [this bug](https://github.com/google/jax/issues/469).

This pr fixes that, and also fixes similar instability in the gradient of logit, for values close to 1. I used the new `custom_transforms` to define the simplified derivatives.

I've also setup batching for the new primitives.

__Edit:__ note I've removed a TODO from lax_scipy_test.py which I think was referring to this instability.

### CODE REVIEW

**Comment by mattjj** [comment]
Nice find!

Would it be possible to use `lax._safe_mul` here instead? That operation is basically a multiply except with the convention that `0 * inf = 0`. The reasoning is that if you have a zero input tangent vector, the result of applying the Jacobian should be zero (since in principle it could have stayed out of the trace entirely).

On the other hand, in general we might expect to need to set up things that are primitive from the point of view of differentiation, but non-primitive from the point of view of other transformations; that's what `custom_transforms` aims to provide!

**Comment by j-towns** [comment]
I’m slightly confused. I think the NAN expit gradient is caused by an inf / inf, not a 0 * inf, so it can’t be fixed by using safe_mul. Maybe I’m wrong. The naive derivative of expit is

exp(-x) / (1 - exp(-x))**2

and I think the issue is caused by both the numerator and the denominator evaluating to inf.

**Comment by mattjj** [comment]
Thanks, I just hadn't thought about it, and I wanted to make you do my thinking for me :) If there's no 0 * inf here then indeed safe_mul can't do anything for us.

Thanks for the clear explanation! That makes it easy for anyone to parse out the reasoning here, including our future selves.

**Review by mattjj — APPROVED** [review]
Great code as always, and moreover a great use of `custom_transforms` that we'll probably use as a pattern in several places in the future.

**Comment by mattjj** [comment]
If you already know of other places we should do this kind of thing, let's make a list and put it in the issue tracker.

**Comment by fehiepsi** [comment]
@neerajprad I think some of your custom primitives `xlogy`, `xlog1py` can follow the same pattern here (so no need to convert impl to jaxpr, write translation rule,...)
